src/execution_plan/execution_plan_reduce_to_apply.c

```diff
@@ -0,0 +1,164 @@
+	match_branch_plan->root = NULL;
+	match_branch_plan->record_map = NULL;
+	// Free the temporary plan.
+	ExecutionPlan_Free(match_branch_plan);
+
+	return branch_match_root;
+}
+
+// Swap operation on operation children.
+static void _OpBaseSwapChildren(OpBase *op, int a, int b) {
+	assert(a >= 0 && b >= 0 && a < op->childCount && b < op->childCount);
+	OpBase *tmp = op->children[a];
+	op->children[a] = op->children[b];
+	op->children[b] = tmp;
+}
+
+static void _CreateBoundBranch(OpBase *op, ExecutionPlan *plan) {
+	rax *bound_vars = NULL;
+	const char **arguments = NULL;
+	if(plan->root) {
+		bound_vars = raxNew();
+		// Rather than cloning the record map, collect the bound variables along with their
+		// parser-generated constant strings.
+		ExecutionPlan_BoundVariables(plan->root, bound_vars);
+		// Prepare the variables for populating the Argument ops we will build.
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+	OpBase *arg = NewArgumentOp(plan, arguments);
+
+	/* In case of Apply operations, we need to append, the new argument branch and
+	 * set it as the first child. */
+	if(op->type & (OPType_APPLY_MULTIPLEXER | OPType_SEMI_APPLY)) {
+		ExecutionPlan_AddOp(op, arg);
+		_OpBaseSwapChildren(op, 0, op->childCount - 1);
+	} else {
+		// -1 & everything is true;
+		op = ExecutionPlan_LocateFirstOp(op, -1);
+		ExecutionPlan_AddOp(op, arg);
+	}
+}
+
+/* This method reduces a filter tree into an OpBase. The method perfrom post-order traversal over the
+ * filter tree, and checks if if the current subtree rooted at the visited node contains path filter or not,
+ * and either reduces the root or continue traversal and reduction.
+ * The three possible operations could be:
+ * 1. OpFilter - In the case the subtree originated from the root does not contains any path filters.
+ * 2. OpSemiApply - In case of the current root is an expression which contains a path filter.
+ * 3. ApplyMultiplexer - In case the current root is an operator (OR or AND) and at least one of its children
+ * has been reduced to OpSemiApply or ApplyMultiplexer. */
+static OpBase *_ExecutionPlan_FilterTreeToOpBaseReduction(ExecutionPlan *plan,
+														  FT_FilterNode *filter_root) {
+	FT_FilterNode *node;
+	// Case of an expression which contains path filter.
+	if(filter_root->t == FT_N_EXP && FilterTree_containsFunc(filter_root, "path_filter", &node)) {
+		AR_ExpNode *expression = filter_root->exp.exp;
+		bool anti = false;
+		if(strcasecmp(expression->op.func_name, "not") == 0) {
+			anti = true;
+			expression = expression->op.children[0];
+		}
+		// Build new Semi Apply op.
+		OpBase *op_semi_apply = NewSemiApplyOp(plan, anti);
+		const cypher_astnode_t *path = expression->op.children[0]->operand.constant.ptrval;
+		// Add a match branch as a Semi Apply op child.
+		ExecutionPlan_AddOp(op_semi_apply, _buildMatchBranch(plan, path));
+		return op_semi_apply;
+	}
+	// Case of an operator (Or or And) which its subtree contains path filter
```

Comment:
I think that any AND tree at this point would have been decomposed into separate Filter ops; we only have OR trees right now. Or maybe that's not true on recursive visits?

Reply:
not valid for ` WHERE n.x = 1 OR (n.x = 2 AND x.y=3)`

Reply:
:+1: